README.rst
==================================================
pyfasta: pythonic access to fasta sequence files.
==================================================


:Author: Brent Pedersen (brentp)
:Email: bpederse@gmail.com
:License: MIT

.. contents ::

Implementation
==============

Requires Python >= 2.6. Stores a flattened version of the fasta file without
spaces or headers and uses either a mmap of numpy binary format or fseek/fread so the
*sequence data is never read into memory*. Saves a pickle (.gdx) of the start, stop 
(for fseek/mmap) locations of each header in the fasta file for internal use.

Usage
=====
::
  
    >>> from pyfasta import Fasta

    >>> f = Fasta('tests/data/three_chrs.fasta')
    >>> sorted(f.keys())
    ['chr1', 'chr2', 'chr3']

    >>> f['chr1']
    NpyFastaRecord(0..80)



Slicing
-------
::

    # get full the sequence:
    >>> a = str(f['chr1'])
    >>> b = f['chr1'][:]
    >>> a == b
    True

    >>> f['chr1'][:10]
    'ACTGACTGAC'

    # get the 1st basepair in every codon (it's python yo)
    >>> f['chr1'][::3]
    'AGTCAGTCAGTCAGTCAGTCAGTCAGT'

    # can query by a 'feature' dictionary (note this is one based coordinates)
    >>> f.sequence({'chr': 'chr1', 'start': 2, 'stop': 9})
    'CTGACTGA'

    # same as:
    >>> f['chr1'][1:9]
    'CTGACTGA'

    # use python, zero based coords
    >>> f.sequence({'chr': 'chr1', 'start': 2, 'stop': 9}, one_based=False)
    'TGACTGA'

    # with reverse complement (automatic for - strand)
    >>> f.sequence({'chr': 'chr1', 'start': 2, 'stop': 9, 'strand': '-'})
    'TCAGTCAG'

Key Function
------------
Sometimes your fasta will have a long header like: "AT1G51370.2 | Symbols:  | F-box family protein | chr1:19045615-19046748 FORWARD" when you only want to key off: "AT1G51370.2". In this case, specify the key_fn argument to the constructor:

::

    >>> fkey = Fasta('tests/data/key.fasta', key_fn=lambda key: key.split()[0])
    >>> sorted(fkey.keys())
    ['a', 'b', 'c']

Numpy
=====

The default is to use a memmaped numpy array as the backend. In which case it's possible to
get back an array directly...
::

    >>> f['chr1'].as_string = False
    >>> f['chr1'][:10] # doctest: +NORMALIZE_WHITESPACE
    memmap(['A', 'C', 'T', 'G', 'A', 'C', 'T', 'G', 'A', 'C'], dtype='|S1')

    >>> import numpy as np
    >>> a = np.array(f['chr2'])
    >>> a.shape[0] == len(f['chr2'])
    True

    >>> a[10:14] # doctest: +NORMALIZE_WHITESPACE
    array(['A', 'A', 'A', 'A'], dtype='|S1')

mask a sub-sequence
::

    >>> a[11:13] = np.array('N', dtype='S1')
    >>> a[10:14].tostring()
    'ANNA'


Backends (Record class)
=======================
It's also possible to specify another record class as the underlying work-horse
for slicing and reading. Currently, there's just the default: 

  * NpyFastaRecord which uses numpy memmap
  * FastaRecord, which uses using fseek/fread
  * MemoryRecord which reads everything into memory and must reparse the original
    fasta every time.
  * TCRecord which is identical to NpyFastaRecord except that it saves the index
    in a TokyoCabinet hash database, for cases when there are enough records that
    loading the entire index from a pickle into memory is unwise. (NOTE: that the
    sequence is not loaded into memory in either case).

It's possible to specify the class used with the `record_class` kwarg to the `Fasta`
constructor:
::

    >>> from pyfasta import FastaRecord # default is NpyFastaRecord
    >>> f = Fasta('tests/data/three_chrs.fasta', record_class=FastaRecord)
    >>> f['chr1']
    FastaRecord('tests/data/three_chrs.fasta.flat', 0..80)

other than the repr, it should behave exactly like the Npy record class backend

it's possible to create your own using a sub-class of FastaRecord. see the source 
in pyfasta/records.py for details.

Flattening
==========
In order to efficiently access the sequence content, pyfasta saves a separate, flattened file with all newlines and headers removed from the sequence. In the case of large fasta files, one may not wish to save 2 copies of a 5GG+ file. In that case, it's possible to flatten the file "inplace", keeping all the headers, and retaining the validity of the fasta file -- with the only change being that the new-lines are removed from each sequence. This can be specified via `flatten_inplace` = True
::
    
    >>> import os
    >>> os.unlink('tests/data/three_chrs.fasta.gdx') # cleanup non-inplace idx
    >>> f = Fasta('tests/data/three_chrs.fasta', flatten_inplace=True)
    >>> f['chr1']  # note the difference in the output from above.
    NpyFastaRecord(6..86)

    # sequence from is same as when requested from non-flat file above.
    >>> f['chr1'][1:9]
    'CTGACTGA'

    # the flattened file is kept as a place holder without the sequence data.
    >>> open('tests/data/three_chrs.fasta.flat').read()
    '@flattened@'


Command Line Interface
======================
there's also a command line interface to manipulate / view fasta files.
the `pyfasta` executable is installed via setuptools, running it will show
help text.

split a fasta file into 6 new files of relatively even size:

  $ pyfasta **split** -n 6 original.fasta

split the fasta file into one new file per header with "%(seqid)s" being filled into each filename.:
  
  $ pyfasta **split** --header "%(seqid)s.fasta" original.fasta

create 1 new fasta file with the sequence split into 10K-mers:

  $ pyfasta **split** -n 1 -k 10000 original.fasta

2 new fasta files with the sequence split into 10K-mers with 2K overlap:

  $ pyfasta **split** -n 2 -k 10000 -o 2000 original.fasta


show some info about the file (and show gc content):

  $ pyfasta **info** --gc test/data/three_chrs.fasta


**extract** sequence from the file. use the header flag to make
a new fasta file. the args are a list of sequences to extract.

  $ pyfasta **extract** --header --fasta test/data/three_chrs.fasta seqa seqb seqc

**extract** sequence from a file using a file containing the headers *not* wanted in the new file:

  $ pyfasta extract --header --fasta input.fasta --exclude --file seqids_to_exclude.txt

**extract** sequence from a fasta file with complex keys where we only want to lookup based on the part before the space.

  $ pyfasta extract --header --fasta input.with.keys.fasta --space --file seqids.txt

**flatten** a file inplace, for faster later use by pyfasta, and without creating another copy. (`Flattening`_)

  $ pyfasta flatten input.fasta 

cleanup 
=======
(though for real use these will remain for faster access)
::

    >>> os.unlink('tests/data/three_chrs.fasta.gdx')
    >>> os.unlink('tests/data/three_chrs.fasta.flat')

Testing
=======
there is currently > 99% test coverage for the 2 modules and all included 
record classes. to run the tests:
::

  $ python setup.py nosetests


CHANGELOG.txt
Changes
=======
0.5.2
-----
fix complement (@mruffalo)

0.5.0
-----
python 3 compatibility thanks to mruffalo

0.4.5
-----
pyfasta split can handle > 52 files. (thanks Devtulya)

0.4.4
-----
fix pyfasta extract

0.4.3
-----
Add 0 or 1-based intervals in sequence() thanks @jamescasbon

0.4.2
-----
update for latest numpy (can't close memmap)

0.4.1
-----
check for duplicate headers.

0.4.0
-----
* add key_fn kwarg to constuctor

0.3.9
-----
* only require 'r' (not r+) for memory map.

0.3.8
-----
* clean up logic for mixing inplace/non-inplace flattened files.
  if the inplace is available, it is always used. 

0.3.6/7
-------
* dont re-flatten the file every time!
* allow spaces before and after the header in the orginal fasta.

0.3.5
-----

* update docs in README.txt for new CLI stuff.
* allow flattening inplace.
* get rid of memmap (results in faster parsing).

0.3.4
-----

* restore python2.5 compatiblity.
* CLI: add ability to exclude sequence from extract
* CLI: allow spliting based on header.

0.3.3
-----

* include this file in the tar ball (thanks wen h.)

0.3.2
-----

* separate out backends into records.py

* use nosetests (python setup.py nosetests)

* add a TCRecord backend for next-gen sequencing availabe if tc is (easy-)installed.

* improve test coverage.


tests.sh
A=`python -c "from pyfasta import Fasta; f = Fasta('tests/data/three_chrs.fasta', flatten_inplace=False); print f['chr1'][:3]"`
B=`python -c "from pyfasta import Fasta; f = Fasta('tests/data/three_chrs.fasta', flatten_inplace=True); print f['chr1'][:3]"`

if [ "$A" != "$B" ]
then
    echo "$A=$B"
fi

F=`cat tests/data/three_chrs.fasta.flat`
if [ "$F" != "@flattened@" ]
then
    echo "$F=@flattened@"
fi


rm -f tests/data/three_chrs.fasta.{flat,gdx}
C=`python -c "from pyfasta import Fasta; f = Fasta('tests/data/three_chrs.fasta', flatten_inplace=True); print f['chr1'][:3]"`
D=`python -c "from pyfasta import Fasta; f = Fasta('tests/data/three_chrs.fasta', flatten_inplace=False); print f['chr1'][:3]"`

if [ "$C" != "$D" ]
then
    echo "$C=$D"
fi

F=`cat tests/data/three_chrs.fasta.flat`
if [ "$F" != "@flattened@" ]
then
    echo "$F=@flattened@"
fi


rm -f tests/data/three_chrs.fasta.{flat,gdx}
I=`python pyfasta/__init__.py info tests/data/three_chrs.fasta`
rm -f tests/data/three_chrs.fasta.{flat,gdx}

# now make sure it works with inplace, and doesnt update.flat
C=`python -c "from pyfasta import Fasta; f = Fasta('tests/data/three_chrs.fasta', flatten_inplace=True); print f['chr1'][:3]"`
J=`python pyfasta/__init__.py info tests/data/three_chrs.fasta`
if [ "$I" != "$J" ]
then
    echo "BADDD"
fi

F=`cat tests/data/three_chrs.fasta.flat`
if [ "$F" != "@flattened@" ]
then
    echo "BADD"
fi

rm -f tests/data/three_chrs.fasta.{flat,gdx}

C=$(python pyfasta/__init__.py extract --fasta tests/data/key.fasta "a extra")
if [ $C != "a" ]
then
    echo "BADD"
fi
rm -f tests/data/key.fasta.{flat,gdx}


tests/bench.py
from __future__ import print_function
from itertools import islice
import sys
import os
sys.path.insert(0, os.path.abspath("."))
import pyfasta
from pyfasta import Fasta

import time
import random
random.seed(1234)

SEQLEN = 100000
def make_long_fasta(filename="t.fasta", nrecs=2500, seqlen=SEQLEN):
    fh = open(filename, 'w')
        #0123456789"
    s = "ACTGACTGAC"
    for i in range(nrecs):
        print(">header%i" % i, file=fh)
        print(s * (seqlen/10), file=fh)

    fh.close()
    return filename

def read(f, nreads=40000, seqlen=SEQLEN):

    for k in islice(f.iterkeys(), 10):
        for i in range(nreads):
            start = random.randint(0, seqlen)
            end = min(seqlen, start + random.randint(1000, 2000))
            str(f[k][start:end])
        


def main():
    fa = make_long_fasta()

    t = time.time()
    f = Fasta(fa)
    print("flatten:", time.time() - t)

    
    t = time.time()
    read(f)
    print("read:", time.time() - t)

     


if __name__ == "__main__":
    main()


tests/test_all.py
from pyfasta import Fasta
from pyfasta.records import NpyFastaRecord, MemoryRecord, FastaRecord
record_classes = [NpyFastaRecord, MemoryRecord, FastaRecord]
from pyfasta import DuplicateHeaderException

try:
    from pyfasta.records import TCRecord
    record_classes.append(TCRecord)
except ImportError:
    pass

import os
import shutil
from nose.tools import assert_raises
import numpy as np
import glob

def _cleanup():
    for f in glob.glob("tests/data/three_chrs.fasta*") + glob.glob('tests/data/dups.fasta.*'):
        if f.endswith(".orig"): continue
        os.unlink(f)
    shutil.copyfile('tests/data/three_chrs.fasta.orig', 'tests/data/three_chrs.fasta')


def test_classes():

    for inplace in (True, False):
        for klass in record_classes:
            f = Fasta('tests/data/three_chrs.fasta', record_class=klass, flatten_inplace=inplace)
            yield check_keys, f
            yield check_misc, f, klass
            yield check_contains, f
            yield check_shape, f
            yield check_bounds, f
            yield check_tostring, f
            yield check_kmers, f
            yield check_kmer_overlap, f
            yield check_slice_size, f
            yield check_slice, f
            yield check_full_slice, f
            yield check_array_copy, f
            yield check_array, f
            yield check_one_based, f

            fasta_name = f.fasta_name

            del f

            yield check_keyfn, 'tests/data/key.fasta', klass, inplace

            yield check_keyfn2, 'tests/data/key.fasta', klass, inplace

            yield check_reload, klass, fasta_name

            yield check_duplicates, klass, inplace

            _cleanup()


def check_duplicates(klass, inplace):
    assert_raises(DuplicateHeaderException,
                  lambda: Fasta('tests/data/dups.fasta',
                                record_class=klass, flatten_inplace=inplace))


def check_keys(f):
    assert sorted(f.keys()) == ['chr1', 'chr2', 'chr3']
    assert sorted(f.iterkeys()) == ['chr1', 'chr2', 'chr3']

def fix(path):
    import os.path as op

    for ext in (".gdx", ".flat"):
        if op.exists(path + ext):
            os.unlink(path + ext)

    shutil.copyfile(path + ".orig", path)

def check_keyfn(path, klass, inplace):
    f = Fasta(path, record_class=klass, flatten_inplace=inplace, key_fn=lambda key: key.split()[0])
    assert sorted(f.keys()) == ['a', 'b', 'c'], f.keys()
    fix(path)
    ff = Fasta(path, record_class=klass, flatten_inplace=inplace)
    assert sorted(ff.keys()) == ['a extra', 'b extra', 'c extra'], (ff.keys(), klass)
    fix(path)

def check_keyfn2(path, klass, inplace):
    f = Fasta(path, record_class=klass, flatten_inplace=inplace, key_fn=lambda
            key: "-".join(key.split()))

    assert sorted(f.keys()) == ['a-extra', 'b-extra', 'c-extra'], f.keys()

    assert f['a-extra']
    fix(path)


def check_reload(klass, fasta_name):

    m = ""
    if klass.__name__ in ('FastaRecord',): # 'NpyFastaRecord'):
        m = os.stat(fasta_name + klass.ext).st_mtime 

    f = Fasta(fasta_name, record_class=klass)
    assert f
    del f
    if klass.__name__ in ('FastaRecord',): # 'NpyFastaRecord'):
        assert os.stat(fasta_name + klass.ext).st_mtime  == m


def check_full_slice(f):
    for k in f.iterkeys():
        assert str(f[k]) == f[k][:]
        assert str(f[k]) == f[k][0:]


        assert str(f[k])[::2] == f[k][0:][::2]
        assert str(f[k])[::2] == f[k][:][::2]

def check_contains(f):

    assert not '_________' in f
    assert 'chr2' in f
    
def check_misc(f, klass):
    seq = f['chr2']
    assert seq.__class__ == klass

    assert (seq[0] == 'T')
    assert (seq[-1] == 'T')

    assert (f['chr3'][0:5][::-1] == 'ACGCA')


    for i in (1, len(seq) -2):
        assert (seq[i] == 'A')

    for i in (1, len(seq) -3):
        assert (seq[i: i + 2] == 'AA')
    
    assert (seq[1] == 'A')

    assert (seq[-1] == 'T')
    assert (seq[-2] == 'A')

    assert (seq[0] == 'T')
    assert (seq[6:9] == 'AAA')

    seq = 'TACGCACGCTAC'
    assert (seq == f['chr3'][-12:])
    assert (seq[:4] == f['chr3'][-12:-8])

def check_shape(f):
    assert (len(f['chr2']) == 80)
    assert (len(f['chr3']) == 3600)
    assert (f['chr2'][:2] == 'TA')

    assert (f['chr3'][:5] == 'ACGCA')

def check_bounds(f):
    c2 = f['chr2']
    assert (len(str(c2[0:900])) == 80)

    assert (c2[99:99] == "")
    assert (c2[99:99] == "")
    assert (c2[80:81] == "")
    assert (c2[79:81] == "T")

    assert (c2[800:810] == "")
    assert_raises(IndexError, lambda: c2[-800])


    assert (c2[-800:-810] == "")

def check_array(f):

    s = f.sequence({'chr': 'chr2', 'start': 1, 'stop':3}, asstring=False)
    assert np.all(s == np.array(['T', 'A', 'A'], dtype="|S1"))


def check_tostring(f):
    s = 'TAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAT'
    assert (str(f['chr2']) == s)

def check_kmers(f):
    seq = str(f['chr2'])

    kmers = list(Fasta.as_kmers(f['chr2'], 10))
    assert (len(kmers) == len(seq) / 10)
    assert (kmers[0] == (0, seq[:10]))

    seqs = [k[1] for k in kmers]
    assert ("".join(seqs) == seq)
    last_pair = kmers[-1]
    assert (seqs[-1][-1] == 'T')

    seq = str(f['chr3'])
    kmers = list(Fasta.as_kmers(f['chr3'], 1))
    assert (kmers[2][0] == 2)
    seqs = [k[1] for k in kmers]
    assert ("".join(seqs) == seq)


def check_kmer_overlap(f):
    chr2 = f['chr2']

    kmers = Fasta.as_kmers(chr2, 10, overlap=2)
    for i, k in enumerate(list(kmers)[:-1]):
        assert (len(k[1]) == 10)
        assert (k[0] == (i * (10 - 2)))

    kmers = Fasta.as_kmers(chr2, 10, overlap=4)
    seqs = [k[1] for k in kmers]
    paired_seqs = zip(seqs[0:-1], seqs[1:])
    for a, b in paired_seqs:
        if len(a) < 4 or len(b) < 4: continue
        assert (a[-4:] == b[:4])

def check_slice_size(f):
    assert (f['chr3'][:7] == 'ACGCATT')
    # take the first basepair of each codon
    assert (f['chr3'][0:7:3] == 'ACT')
    # take the 2nd basepair of each codon
    assert (f['chr3'][1:7:3] == 'CA')

def check_slice(f):
    assert (str(f['chr3'][:4]) == 'ACGC')


    seq = 'TACGCACGCTAC'
    assert (seq == f['chr3'][-12:])

    assert (seq[:4] == f['chr3'][-12:-8])

    assert (f['chr3'][0:5][::-1] == 'ACGCA')
    assert (f['chr3'][0:5] == 'ACGCA')


def check_array_copy(f):
    # the array is definitely a copy...
    a = np.array(f['chr3'], np.dtype('S1'))
    old = f['chr3'][1:5]
    assert (a.shape[0] == 3600)
    a[1:5] = np.array('N', dtype='S1')
    c = f['chr3'][1:5]
    assert c == old

    assert a[1:5].tostring().decode() == 'NNNN', a[1:5].tostring().decode()

def check_one_based(f):
    assert f.sequence({'chr': 'chr1', 'start': 2, 'stop': 9})  == 'CTGACTGA'
    assert f.sequence({'chr': 'chr1', 'start': 2, 'stop': 9}, one_based=False) == 'TGACTGA'

if __name__ == "__main__":
    import nose
    nose.main()


ez_setup.py
#!python
"""Bootstrap setuptools installation

If you want to use setuptools in your package's setup.py, just include this
file in the same directory with it, and add this to the top of your setup.py::

    from ez_setup import use_setuptools
    use_setuptools()

If you want to require a specific version of setuptools, set a download
mirror, or use an alternate download directory, you can do so by supplying
the appropriate options to ``use_setuptools()``.

This file can also be run as a script to install or upgrade setuptools.
"""
from __future__ import print_function
import sys
DEFAULT_VERSION = "0.6c11"
DEFAULT_URL     = "http://pypi.python.org/packages/%s/s/setuptools/" % sys.version[:3]

md5_data = {
    'setuptools-0.6b1-py2.3.egg': '8822caf901250d848b996b7f25c6e6ca',
    'setuptools-0.6b1-py2.4.egg': 'b79a8a403e4502fbb85ee3f1941735cb',
    'setuptools-0.6b2-py2.3.egg': '5657759d8a6d8fc44070a9d07272d99b',
    'setuptools-0.6b2-py2.4.egg': '4996a8d169d2be661fa32a6e52e4f82a',
    'setuptools-0.6b3-py2.3.egg': 'bb31c0fc7399a63579975cad9f5a0618',
    'setuptools-0.6b3-py2.4.egg': '38a8c6b3d6ecd22247f179f7da669fac',
    'setuptools-0.6b4-py2.3.egg': '62045a24ed4e1ebc77fe039aa4e6f7e5',
    'setuptools-0.6b4-py2.4.egg': '4cb2a185d228dacffb2d17f103b3b1c4',
    'setuptools-0.6c1-py2.3.egg': 'b3f2b5539d65cb7f74ad79127f1a908c',
    'setuptools-0.6c1-py2.4.egg': 'b45adeda0667d2d2ffe14009364f2a4b',
    'setuptools-0.6c10-py2.3.egg': 'ce1e2ab5d3a0256456d9fc13800a7090',
    'setuptools-0.6c10-py2.4.egg': '57d6d9d6e9b80772c59a53a8433a5dd4',
    'setuptools-0.6c10-py2.5.egg': 'de46ac8b1c97c895572e5e8596aeb8c7',
    'setuptools-0.6c10-py2.6.egg': '58ea40aef06da02ce641495523a0b7f5',
    'setuptools-0.6c11-py2.3.egg': '2baeac6e13d414a9d28e7ba5b5a596de',
    'setuptools-0.6c11-py2.4.egg': 'bd639f9b0eac4c42497034dec2ec0c2b',
    'setuptools-0.6c11-py2.5.egg': '64c94f3bf7a72a13ec83e0b24f2749b2',
    'setuptools-0.6c11-py2.6.egg': 'bfa92100bd772d5a213eedd356d64086',
    'setuptools-0.6c2-py2.3.egg': 'f0064bf6aa2b7d0f3ba0b43f20817c27',
    'setuptools-0.6c2-py2.4.egg': '616192eec35f47e8ea16cd6a122b7277',
    'setuptools-0.6c3-py2.3.egg': 'f181fa125dfe85a259c9cd6f1d7b78fa',
    'setuptools-0.6c3-py2.4.egg': 'e0ed74682c998bfb73bf803a50e7b71e',
    'setuptools-0.6c3-py2.5.egg': 'abef16fdd61955514841c7c6bd98965e',
    'setuptools-0.6c4-py2.3.egg': 'b0b9131acab32022bfac7f44c5d7971f',
    'setuptools-0.6c4-py2.4.egg': '2a1f9656d4fbf3c97bf946c0a124e6e2',
    'setuptools-0.6c4-py2.5.egg': '8f5a052e32cdb9c72bcf4b5526f28afc',
    'setuptools-0.6c5-py2.3.egg': 'ee9fd80965da04f2f3e6b3576e9d8167',
    'setuptools-0.6c5-py2.4.egg': 'afe2adf1c01701ee841761f5bcd8aa64',
    'setuptools-0.6c5-py2.5.egg': 'a8d3f61494ccaa8714dfed37bccd3d5d',
    'setuptools-0.6c6-py2.3.egg': '35686b78116a668847237b69d549ec20',
    'setuptools-0.6c6-py2.4.egg': '3c56af57be3225019260a644430065ab',
    'setuptools-0.6c6-py2.5.egg': 'b2f8a7520709a5b34f80946de5f02f53',
    'setuptools-0.6c7-py2.3.egg': '209fdf9adc3a615e5115b725658e13e2',
    'setuptools-0.6c7-py2.4.egg': '5a8f954807d46a0fb67cf1f26c55a82e',
    'setuptools-0.6c7-py2.5.egg': '45d2ad28f9750e7434111fde831e8372',
    'setuptools-0.6c8-py2.3.egg': '50759d29b349db8cfd807ba8303f1902',
    'setuptools-0.6c8-py2.4.egg': 'cba38d74f7d483c06e9daa6070cce6de',
    'setuptools-0.6c8-py2.5.egg': '1721747ee329dc150590a58b3e1ac95b',
    'setuptools-0.6c9-py2.3.egg': 'a83c4020414807b496e4cfbe08507c03',
    'setuptools-0.6c9-py2.4.egg': '260a2be2e5388d66bdaee06abec6342a',
    'setuptools-0.6c9-py2.5.egg': 'fe67c3e5a17b12c0e7c541b7ea43a8e6',
    'setuptools-0.6c9-py2.6.egg': 'ca37b1ff16fa2ede6e19383e7b59245a',
}

import os
from hashlib import md5

def _validate_md5(egg_name, data):
    if egg_name in md5_data:
        digest = md5(data).hexdigest()
        if digest != md5_data[egg_name]:
            print((
                "md5 validation of %s failed!  (Possible download problem?)"
                % egg_name
            ), file=sys.stderr)
            sys.exit(2)
    return data

def use_setuptools(
    version=DEFAULT_VERSION, download_base=DEFAULT_URL, to_dir=os.curdir,
    download_delay=15
):
    """Automatically find/download setuptools and make it available on sys.path

    `version` should be a valid setuptools version number that is available
    as an egg for download under the `download_base` URL (which should end with
    a '/').  `to_dir` is the directory where setuptools will be downloaded, if
    it is not already available.  If `download_delay` is specified, it should
    be the number of seconds that will be paused before initiating a download,
    should one be required.  If an older version of setuptools is installed,
    this routine will print a message to ``sys.stderr`` and raise SystemExit in
    an attempt to abort the calling script.
    """
    was_imported = 'pkg_resources' in sys.modules or 'setuptools' in sys.modules
    def do_download():
        egg = download_setuptools(version, download_base, to_dir, download_delay)
        sys.path.insert(0, egg)
        import setuptools; setuptools.bootstrap_install_from = egg
    try:
        import pkg_resources
    except ImportError:
        return do_download()
    try:
        pkg_resources.require("setuptools>="+version); return
    except pkg_resources.VersionConflict as e:
        if was_imported:
            print((
            "The required version of setuptools (>=%s) is not available, and\n"
            "can't be installed while this script is running. Please install\n"
            " a more recent version first, using 'easy_install -U setuptools'."
            "\n\n(Currently using %r)"
            ) % (version, e.args[0]), file=sys.stderr)
            sys.exit(2)
        else:
            del pkg_resources, sys.modules['pkg_resources']    # reload ok
            return do_download()
    except pkg_resources.DistributionNotFound:
        return do_download()

def download_setuptools(
    version=DEFAULT_VERSION, download_base=DEFAULT_URL, to_dir=os.curdir,
    delay = 15
):
    """Download setuptools from a specified location and return its filename

    `version` should be a valid setuptools version number that is available
    as an egg for download under the `download_base` URL (which should end
    with a '/'). `to_dir` is the directory where the egg will be downloaded.
    `delay` is the number of seconds to pause before an actual download attempt.
    """
    import urllib2
    egg_name = "setuptools-%s-py%s.egg" % (version,sys.version[:3])
    url = download_base + egg_name
    saveto = os.path.join(to_dir, egg_name)
    src = None
    if not os.path.exists(saveto):  # Avoid repeated downloads
        try:
            from distutils import log
            if delay:
                log.warn("""
---------------------------------------------------------------------------
This script requires setuptools version %s to run (even to display
help).  I will attempt to download it for you (from
%s), but
you may need to enable firewall access for this script first.
I will start the download in %d seconds.

(Note: if this machine does not have network access, please obtain the file

   %s

and place it in this directory before rerunning this script.)
---------------------------------------------------------------------------""",
                    version, download_base, delay, url
                ); from time import sleep; sleep(delay)
            log.warn("Downloading %s", url)
            src = urllib2.urlopen(url)
            # Read/write all in one block, so we don't create a corrupt file
            # if the download is interrupted.
            data = _validate_md5(egg_name, src.read())
            with open(saveto, "wb") as dst:
                dst.write(data)
        finally:
            if src: src.close()
    return os.path.realpath(saveto)




































def main(argv, version=DEFAULT_VERSION):
    """Install or upgrade setuptools and EasyInstall"""
    try:
        import setuptools
    except ImportError:
        egg = None
        try:
            egg = download_setuptools(version, delay=0)
            sys.path.insert(0,egg)
            from setuptools.command.easy_install import main
            return main(list(argv)+[egg])   # we're done here
        finally:
            if egg and os.path.exists(egg):
                os.unlink(egg)
    else:
        if setuptools.__version__ == '0.0.1':
            print((
            "You have an obsolete version of setuptools installed.  Please\n"
            "remove it from your system entirely before rerunning this script."
            ), file=sys.stderr)
            sys.exit(2)

    req = "setuptools>="+version
    import pkg_resources
    try:
        pkg_resources.require(req)
    except pkg_resources.VersionConflict:
        try:
            from setuptools.command.easy_install import main
        except ImportError:
            from easy_install import main
        main(list(argv)+[download_setuptools(delay=0)])
        sys.exit(0) # try to force an exit
    else:
        if argv:
            from setuptools.command.easy_install import main
            main(argv)
        else:
            print("Setuptools version",version,"or greater has been installed.")
            print('(Run "ez_setup.py -U setuptools" to reinstall or upgrade.)')

def update_md5(filenames):
    """Update our built-in md5 registry"""

    import re

    for name in filenames:
        base = os.path.basename(name)
        with open(name, 'rb') as f:
            md5_data[base] = md5(f.read()).hexdigest()

    data = ["    %r: %r,\n" % it for it in md5_data.iteritems()]
    data.sort()
    repl = "".join(data)

    import inspect
    srcfile = inspect.getsourcefile(sys.modules[__name__])
    with open(srcfile) as f:
        src = f.read()

    match = re.search("\nmd5_data = {\n([^}]+)}", src)
    if not match:
        print("Internal error!", file=sys.stderr)
        sys.exit(2)

    src = src[:match.start(1)] + repl + src[match.end(1):]
    with open(srcfile, 'w') as f:
        f.write(src)


if __name__=='__main__':
    if len(sys.argv)>2 and sys.argv[1]=='--md5update':
        update_md5(sys.argv[2:])
    else:
        main(sys.argv[1:])








pyfasta/__init__.py
from __future__ import print_function
import sys
from fasta import Fasta, complement, DuplicateHeaderException
from records import *
from split_fasta import split
import optparse

def main():
    help = """
    available actions:
        `extract`: extract sequences from a fasta file
        `info`: show info about the fasta file and exit.
        `split`: split a large fasta file into separate files
                 and/or into K-mers.
        `flatten`: flatten a fasta file inplace so that later
                   command-line (and programmattic) access via
                   pyfasta will use the inplace flattened version
                   rather than creating another .flat copy of the
                   sequence.

    to view the help for a particular action, use:
        pyfasta [action] --help
    e.g.:
        pyfasta extract --help
    """
    if len(sys.argv) == 1:
        print(help)
        sys.exit()

    action = sys.argv[1]

    sglobals = globals()
    if not action in sglobals:
        print("%s not a valid action" % action)
        print(help)
        sys.exit()

    globals()[action](sys.argv[2:])

def info(args):
    """
    >>> info(['tests/data/three_chrs.fasta'])
    <BLANKLINE>
    tests/data/three_chrs.fasta
    ===========================
    >chr3 length:3600
    >chr2 length:80
    >chr1 length:80
    <BLANKLINE>
    3760 basepairs in 3 sequences
    """
    parser = optparse.OptionParser("""\
   print headers and lengths of the given fasta file in order of length. e.g.:
        pyfasta info --gc some.fasta""")

    parser.add_option("-n", "--n", type="int", dest="nseqs",
                      help="max number of records to print. use -1 for all",
                      default=20)
    parser.add_option("--gc", dest="gc", help="show gc content",
                      action="store_true", default=False)
    options, fastas = parser.parse_args(args)
    if not (fastas):
        sys.exit(parser.print_help())
    import operator

    for fasta in fastas:
        f = Fasta(fasta)
        info = [(k, len(seq)) for k, seq in f.iteritems()]

        total_len = sum(l for k, l in info)
        nseqs = len(f)
        if options.nseqs > -1:
            info = sorted(info,  key=operator.itemgetter(1, 0), reverse=True)
            info = info[:options.nseqs]
        else:
            info.sort()

        print("\n" + fasta)
        print("=" * len(fasta))
        for k, l in info:
            gc = ""
            if options.gc:
                seq = str(f[k]).upper()
                g = seq.count('G')
                c = seq.count('C')
                gc = 100.0 * (g + c) / float(l)
                gc = "gc:%.2f%%" % gc
            print((">%s length:%i" % (k, l)) + gc)

        if total_len > 1000000:
            total_len = "%.3fM" % (total_len / 1000000.)
        print()
        print("%s basepairs in %i sequences" % (total_len, nseqs))

def flatten(args):
    """
    >>> flatten(['tests/data/three_chrs.fasta'])
    """
    parser = optparse.OptionParser("""flatten a fasta file *inplace* so all later access by pyfasta will use that flattend (but still viable) fasta file""")
    _, fasta = parser.parse_args(args)
    for fa in fasta:
        f = Fasta(fa, flatten_inplace=True)

def extract(args):
    """
    >>> extract(['--fasta', 'tests/data/three_chrs.fasta', 'chr2'])
    TAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAT
    """

    parser = optparse.OptionParser("""extract some sequences from a fasta file. e.g.:
               pyfasta extract --fasta some.fasta --header at2g26540 at3g45640""")
    parser.add_option("--fasta", dest="fasta", help="path to the fasta file")
    parser.add_option("--header", dest="header", help="include headers", action="store_true", default=False)
    parser.add_option("--exclude", dest="exclude", help="extract all sequences EXCEPT those listed", action="store_true", default=False)
    parser.add_option("--file", dest="file", help=\
                      "if this flag is used, the sequences to extract" \
                      " are read from the file specified in args"
                      , action="store_true", default=False)
    parser.add_option("--space", dest="space", action="store_true", help=\
                      "use the fasta identifier only up to the space as the key",
                      default=False)
    options, seqs = parser.parse_args(args)
    if not (options.fasta and len(seqs)):
        sys.exit(parser.print_help())

    key_fn = (lambda k: k.split()[0]) if options.space else None
    f = Fasta(options.fasta, key_fn=key_fn)
    if options.file:
        seqs = (x.strip() for x in open(seqs[0]))
    if options.exclude:
        seqs = sorted(frozenset(f.iterkeys()).difference(seqs))

    for seqname in seqs:
        seq = f[seqname]
        if options.header:
            print(">%s" % seqname)
        print(seq)


if __name__ == "__main__":
    main()
